blackjax/adaptation/meads.py

```diff
@@ -0,0 +1,125 @@
+"""Public API for Maximum-Eigenvalue Adaptation of Damping and Step-size kernel"""
+
+from typing import Callable
+
+import jax
+import jax.numpy as jnp
+
+import blackjax.adaptation.chain_adaptation as chain_adaptation
+from blackjax.types import PyTree
+
+
+def base(
+    kernel_factory: Callable,
+    logprob_grad_fn: Callable,
+    batch_fn: Callable = jax.vmap,
+):
+    """Maximum-Eigenvalue Adaptation of Damping and Step size for the Generalized
+    Hamiltonian Monte Carlo kernel [1]_.
+
+    Performs a cross-chain adaptation scheme for the Generalized HMC algorithm that
+    automatically selects values for the Generalized HMC's tunable parameters
+    based on statistics collected from a population of many chains. It uses heuristics
+    determined by the maximum eigenvalue of the covariance and gradient matrices given
+    by the grouped samples of all chains with shape (num_chain, parameter dimension).
+    Specifically, it does Algorithm 3 of [1]_ with cross-chain adaptation instead of
+    parallel ensample chain adaptation.
+
+    Parameters
+    ----------
+    kernel_factory
+        Function that takes as input the step size, alpha and delta parameters
+        and outputs a Generalized HMC kernel that generates new samples.
+    logprob_grad_fn
+        The gradient of logprob_fn, outputs the gradient PyTree for sample.
+    batch_fn
+        Either jax.vmap or jax.pmap to perform parallel operations.
+
+    Returns
+    -------
+    init
+        Function that initializes the warmup.
+    update
+        Function that moves the warmup one step.
+    final
+        Function that returns step size, alpha and delta given a cross-chain
+        warmup state.
+
+    References
+    ----------
+    .. [1]: Hoffman, M. D., & Sountsov, P. (2022). Tuning-Free Generalized
+            Hamiltonian Monte Carlo. In International Conference on Artificial
+            Intelligence and Statistics (pp. 7799-7813). PML
+    """
+
+    def maximum_eigenvalue(matrix: PyTree):
+        """Estimate largest eigenvalues of a matrix.
+
+        Calculates an unbiased estimate of the ratio between the sum of the squared
+        eigenvalues and the sum of the eigenvalues from the input matrix. This ratio
+        approximates the largest eigenvalue well except in cases when there are a
+        large number of small eigenvalues significantly larger than 0 but significantly
+        smaller than the largest eigenvalue. This unbiased estimate is used instead of
+        directly computing an unbiased estimate of the largest eigenvalue because of
+        the latter's large variance.
+        """
+
+        X = jnp.vstack(
+            [
+                leaf.reshape(leaf.shape[0], -1).T
+                for leaf in jax.tree_util.tree_leaves(matrix)
+            ]
+        ).T
```

Comment:
```suggestion
        X = jnp.hstack(
            [
                leaf.reshape(leaf.shape[0], -1)
                for leaf in jax.tree_util.tree_leaves(matrix)
            ]
        )
```

Or add a kwarg of unravel function, where user can pass in function to unravel (which could be `vmapping` a unravel func like in
https://github.com/blackjax-devs/blackjax/blob/main/blackjax/vi/pathfinder.py#L176-L184)